### DISCUSSION TITLE
Welcome to mrsimulator Discussions!

### DISCUSSION THREAD

**Discussion by deepanshs** [deepanshs/mrsimulator#46] [Announcement]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by Max-Kimball** [comment]
Hi, I'm trying to fit a spectrum with multiple sites and J-couplings, I'm trying to figure it out with mrsimulator, but it's my first time playing with code. I'm trying to follow the fitting tutorial, but with multiple spin systems and couplings. I have no problems up until I try to minimize the fit, then I get 

C;\\.....\\anaconda3\lib\site-packages\mrsimulator\utils\spectral_fitting.py", line 447, in LMFIT_min_function
    exp_data = mth.experiment.y[0].components[0]

AttributeError: 'NoneType' object has no attribute 'y'

Do you have any idea what might be causing this?

I can do  everything on the app but add J-coupling. Is this something I'm missing? Or is it not implemented yet? 

Thank you so much for your hard work! It's very impressive and very helpful.

**Reply by deepanshs** [reply]
Hi Max. For fitting, you need to add the measured data to method object, for example (See full example on [readthedocs](https://mrsimulator.readthedocs.io/en/stable/fitting/1D_fitting/plot_5_119Sn_sideband.html#sphx-glr-fitting-1d-fitting-plot-5-119sn-sideband-py))
```py
MAS = BlochDecaySpectrum
    channels=["119Sn"],
    magnetic_flux_density=9.395,  # in T
    rotor_frequency=10000,  # in Hz
    spectral_dimensions=spectral_dims,
    experiment=experiment,  # add the measurement to the method.
)
```
The error you see is from the missing experiment data.

---

The J-couplings are implement in v0.7. See the [isotopomers](https://mrsimulator.readthedocs.io/en/stable/introduction/isotopomers_example.html) example. To add coupling between two sites, use the Coupling class
```py
CH3_coupling_1 = Coupling(site_index=[0, 6], isotropic_j=125)
```
where `isotropic_j` is the isotopic J-coupling.

**Reply by Max-Kimball** [reply]
Ah, I missed the experiment=experiment line. Thank you for the fast reply!

**Comment by Max-Kimball** [comment]
Hello again deepanshs,

I'm producing fits which is very exciting, but they're not very good with the limited processing I'm able to do via python. I have to do some baseline correction and first order phasing.

I managed to trick NMRglue to import my processed spectrum from topspin by saving it as an FID, but of course it's not in the correct units, so I can't move forward. I'm trying to read more of the documentation to figure it out, but I'm a bit out of my depth.

Is there a proper way to bring in the processed spectrum, or to correct the units so I can move on to fitting? Or a way to do the baseline correction and first order phasing?

I'm incredibly thankful for your work, it's very impressive. I'm trying to learn more about python to properly implement it.

**Reply by deepanshs** [reply]
I appreciate you using mrsimulator. Feel free to ask any questions. We'll be happy to assist.

We have an [example](https://mrsimulator.readthedocs.io/en/stable/introduction/fitting_example.html) that shows fitting starting from importing data/signal processing (zero and first-order phase correction). I would recommend following this example first.

---

If you follow the steps in the above example, you should not need to change the dimension unit.
Yes, you can change the [dimension unit](https://csdmpy.readthedocs.io/en/stable/startFromScratch/dimension/generate/dimension_objects.html) of a csdm object by multiplying it by appropriate units. For example, if `csdm_ds` has a unitless dimension, you can convert it to a time dimension by multiplying it by a time unit,
```py
import csdmpy as cp
csdm_ds.dimensions[0] *= cp.ScalarQuantity("s")
```

---

We have a few [baseline functions](https://mrsimulator.readthedocs.io/en/stable/signal_processor/index.html) that you can use for fitting.
